glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt

```diff
@@ -173,9 +85,33 @@ class PingUploadWorker(context: Context, params: WorkerParameters) : Worker(cont
          * @return true if process was successful
          */
         internal fun uploadPings(): Boolean {
-            val storageDirectory = File(Glean.getDataDir(), PINGS_DIR)
+            var ffiTask = LibGleanFFI.INSTANCE.glean_get_upload_task()
+            var task = ffiTask.toPingUploadTask()
+            while (task is PingUploadTask.Upload) {
+                // Get the request
+                val request = task.request()!!
+                // Upload the request
+                // If the status is `null` there was some kind of unrecoverable error
+                // so we return 400 which will ensure this gets treated as such.
+                @Suppress("MagicNumber")
+                val status = Glean.httpClient.doUpload(
+                        request.path,
+                        request.body,
+                        request.headers,
+                        Glean.configuration
+                ) ?: 400
+                // Process the upload response
+                LibGleanFFI.INSTANCE.glean_process_ping_upload_response(ffiTask, status)
```

Comment:
We report back the whole task so that rust can properly drop it. If we don't we have to rely on each platform properly dropping the whole request, but this way the platform is required to give the task back and rust **always** properly drops it.

Reply:
I'm not sure I understand. [Internally](https://github.com/mozilla/glean/blob/2bfcba5dae018bbff3a0b4768cf78d7eab645c67/glean-core/ffi/src/lib.rs#L345), we only really use the uuid and the status. Why do we need the whole `ffiTask`?

Reply:
`ffiTask` contains allocated data (the strings for uuid/headers/...). These need to be freed on the Rust side.
For other strings we do this [when copying out the string](https://github.com/mozilla/glean/blob/master/glean-core/android/src/main/java/mozilla/telemetry/glean/rust/LibGleanFFI.kt#L28-L33).
In here we only copy out the string but don't free its memory (as [getRustString](https://github.com/mozilla/glean/blob/master/glean-core/android/src/main/java/mozilla/telemetry/glean/rust/LibGleanFFI.kt#L40-L42) is used)

By passing back in the whole `ffiTask` the Rust side gets ownership back and when dropped (when the function scope ends) the data gets freed ([here](https://github.com/brizental/glean/blob/1605078/android-swap/glean-core/ffi/src/upload.rs#L48-L66)).

Reply:
Advantage of passing it back and forth is that only the Rust side is responsible for clearing it all (if we ever need to extend it we need to adjust `drop` there).
The platform side doesn't have to care about freeing at all (it _does_ need to be careful to only read the strings, not free them, though)

Reply:
Really just to free the pointers properly, see here: https://github.com/mozilla/glean/blob/2bfcba5dae018bbff3a0b4768cf78d7eab645c67/glean-core/ffi/src/upload.rs#L48-L66. 

We implement Drop for FfiPingUploadTask, which implicity is called once `task` goes out of scope in [this function](https://github.com/mozilla/glean/blob/2bfcba5dae018bbff3a0b4768cf78d7eab645c67/glean-core/ffi/src/lib.rs#L336). For that to happen we need to pass the whole task back.

Reply:
Can you add a tl;dr of this explanation somewhere in the call chain?